src/sc2api/sc2_client.cc

```diff
@@ -2035,73 +2030,69 @@ void ControlImp::IssueUnitDestroyedEvents() {
 }
 
 void ControlImp::IssueUnitAddedEvents() {
-    observation_imp_->unit_pool_.ForEachExistingUnit([&](sc2::Unit& unit) {
-        auto found = observation_imp_->units_previous_map_.find(unit.tag);
-        if (found != observation_imp_->units_previous_map_.end()) {
-            return;
-        }
 
-        if (unit.alliance == Unit::Alliance::Enemy && unit.display_type == Unit::DisplayType::Visible) {
-            client_.OnUnitEnterVision(&unit);
+    for (auto unit : observation_imp_->unit_pool_.GetNewUnits())
+    {
+        if (unit->alliance == Unit::Alliance::Self) {
+            client_.OnUnitCreated(unit);
         }
-        else if (unit.alliance == Unit::Alliance::Self) {
-            client_.OnUnitCreated(&unit);
+        else if (unit->alliance == Unit::Alliance::Neutral && unit->display_type == Unit::DisplayType::Visible) {
+            client_.OnNeutralUnitCreated(unit);
         }
-    });
-}
-
-void ControlImp::IssueIdleEvent(const Unit* unit, const std::vector<Tag>& commands) {
-    if (!unit || !unit->orders.empty() || unit->build_progress < 1.0f) {
-        return;
     }
-    // Lookup unit from previous map.
-    auto found = observation_imp_->units_previous_map_.find(unit->tag);
 
-    // If it's not in the previous map it's a new unit with new orders so trigger the OnIdle event.
-    if (found == observation_imp_->units_previous_map_.end()) {
-        client_.OnUnitIdle(unit);
-        return;
+    for (auto unit : observation_imp_->unit_pool_.GetUnitsEnteringVision())
+    {
+        if (unit->alliance == Unit::Alliance::Enemy && unit->display_type == Unit::DisplayType::Visible) {
+            client_.OnUnitEnterVision(unit);
```

Comment:
Style:
You can safely drop brackets here, i.e.
```cpp
if (unit->alliance == Unit::Alliance::Enemy && unit->display_type == Unit::DisplayType::Visible)
             client_.OnUnitEnterVision(unit);
```

But it is ok to leave it as is.